### PR TITLE
Create a shared library containing AZ::Environment

### PR DESCRIPTION
**Pull request by burelc-amzn**
This change makes a shared library that contains the AZ::Environment singleton.

This is a paradigm shift for how the Environment instance is retrieved. It gets rid of these methods of `Environment`:
* `Attach`
* `Detach`
* `Create`
* `Destroy`
* `IsReady`

The first time that `Environment::GetInstance()` is called, the `Environment` instance is created. It then persists for the duration of the program.

With this change, the `InitializeDynamicModule()` entry point of every gem no longer needs to call `AZ::Environment::Attach()`. Just call `GetInstance()` and the correct Environment instance will be retrieved.

### CODE REVIEW

**Comment by AMZN-alexpete** [comment]
This seems like a nice solution to a thorny problem.  Are there future plans to use this library for other functionality? If not I recommend calling it `AzEnvironment` instead of `O3DEKernel`.

Also, how are we ensuring this library is loaded before it is needed? In glancing through, I didn't see how it was getting loaded in a non-monolithic build.   Should it be added to the runtime dependencies in .cmake anywhere?

**Comment by burelc-amzn** [comment]
> This seems like a nice solution to a thorny problem. Are there future plans to use this library for other functionality? If not I recommend calling it `AzEnvironment` instead of `O3DEKernel`.
@AMZN-alexpete Yes, the `O3DEKernel` library will also contain the allocator instances, which is outlined in the [Allocator Availability RFC](https://github.com/o3de/sig-core/blob/main/rfcs/allocator-availability.md).

> Also, how are we ensuring this library is loaded before it is needed? In glancing through, I didn't see how it was getting loaded in a non-monolithic build. Should it be added to the runtime dependencies in .cmake anywhere?
This library is a typical shared library. The dependency is defined at compile time, and is embedded in the `.dll`/`.so`/`.dylib`. The OS loads the dependent libraries for us at load time. It is just like the way the Editor.exe loads the Qt dlls.

File: Gems/Maestro/Code/Source/Cinematics/AnimPostFXNode.cpp [hunk 1]
```diff
@@ -38,7 +38,7 @@ class CFXNodeDescription
         virtual void GetDefault(bool& val) const = 0;
         virtual void GetDefault(Vec4& val) const = 0;
 
-        AZStd::string m_name;
+        std::string m_name;
```
Comment (amzn-phist) [inline Gems/Maestro/Code/Source/Cinematics/AnimPostFXNode.cpp:41]: What is the motivation for this change?

File: Code/Tools/SceneAPI/SceneCore/SceneCoreStandaloneAllocator.cpp [hunk 2]
```diff
@@ -16,9 +16,8 @@ namespace AZ
     {
         bool SceneCoreStandaloneAllocator::m_allocatorInitialized = false;
 
-        void SceneCoreStandaloneAllocator::Initialize(AZ::EnvironmentInstance environment)
+        void SceneCoreStandaloneAllocator::Initialize(AZ::EnvironmentInstance /*environment*/)
```
Comment (amzn-phist) [inline Code/Tools/SceneAPI/SceneCore/SceneCoreStandaloneAllocator.cpp:19]: This parameter can be removed if it's not ever used.

File: Code/Tools/SceneAPI/SceneData/SceneDataStandaloneAllocator.cpp [hunk 3]
```diff
@@ -16,9 +16,8 @@ namespace AZ
     {
         bool SceneDataStandaloneAllocator::m_allocatorInitialized = false;
 
-        void SceneDataStandaloneAllocator::Initialize(AZ::EnvironmentInstance environment)
+        void SceneDataStandaloneAllocator::Initialize(AZ::EnvironmentInstance /*environment*/)
```
Comment (amzn-phist) [inline Code/Tools/SceneAPI/SceneData/SceneDataStandaloneAllocator.cpp:19]: Same with this one.

File: Code/Framework/AzCore/AzCore/O3DEKernelConfiguration.h [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <AzCore/PlatformDef.h>
+
+#if defined(AZ_MONOLITHIC_BUILD)
+    #define O3DEKERNEL_API
+#else
+    #if defined(O3DEKernel_EXPORTS)
```
Comment (amzn-phist) [inline Code/Framework/AzCore/AzCore/O3DEKernelConfiguration.h:16]: nit: Preprocessor names should be all CAPS.

(Maybe this was discussed before and approved)

File: Code/Framework/AzCore/Tests/DLLMainTest.cpp [hunk 5]
```diff
@@ -105,9 +105,8 @@ extern "C" AZ_DLL_EXPORT void DestroyDLLTestVirtualClass()
 //////////////////////////////////////////////////////////////////////////
 
 
-extern "C" AZ_DLL_EXPORT void InitializeDynamicModule(void* azEnvironmentInstance)
+extern "C" AZ_DLL_EXPORT void InitializeDynamicModule([[maybe_unused]] void* azEnvironmentInstance)
 {
-    AZ::Environment::Attach(static_cast<AZ::EnvironmentInstance>(azEnvironmentInstance));
     AZ_Assert(AZ::Environment::GetInstance() == azEnvironmentInstance, "Invalid!");
     AZ_Printf("DLL", "InitializeDynamicModule called");
```
Comment (amzn-phist) [inline Code/Framework/AzCore/Tests/DLLMainTest.cpp:111]: I'd consider removing the `void*` parameter of `InitializeDynamicModule` altogether.  Seems like this is the only place where it's actually used.  I'd ask whether keeping this Assert gives enough value to us that we opt to keep the parameter.

Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/Tests/DLLMainTest.cpp:111]: The parameter isn't unused. 
It is what is `AZ::EnironmentInstance*` that gets attached to the AZ::Environment.

Comment (burelc-amzn) [reply Code/Framework/AzCore/Tests/DLLMainTest.cpp:111]: My concern here is backwards compatibility. Any custom gem that isn't using the `AZ_DECLARE_MODULE_INITIALIZATION` macro, but rather a custom definition, would break if the parameters are changed.

File: Gems/Maestro/Code/Source/Cinematics/AnimPostFXNode.cpp (hunk 1 above)
Comment (burelc-amzn) [reply Gems/Maestro/Code/Source/Cinematics/AnimPostFXNode.cpp:41]: The commit message for commit 6a9430533 has the details:

>This code uses global static instances, which have a nondeterministic
initialization order. Due to the fact that they are global, the AZ
allocators are not guaranteed to be initialized first. To deal with
this, most of this code already uses `std` containers over `AZStd` ones.
It stores a `CAnimPostFXNode::FxNodeDescriptionMap`, which is a
`std::map`, the values of which are `CFXNodeDescription`s. Those
descriptions were storing an `AZStd::string`. Putting an `AZStd::string`
in a global static variable is not guaranteed to work.

File: Code/Framework/AzCore/AzCore/O3DEKernelConfiguration.h (hunk 4 above)
Comment (burelc-amzn) [reply Code/Framework/AzCore/AzCore/O3DEKernelConfiguration.h:16]: This is the default value that CMake creates for us, based on the [DEFINE_SYMBOL property](https://cmake.org/cmake/help/latest/prop_tgt/DEFINE_SYMBOL.html). We can change it, but do we need to? It won't appear anywhere else except in this header.

File: Code/Framework/AzCore/Tests/DLLMainTest.cpp (hunk 5 above)
Comment (amzn-phist) [reply Code/Framework/AzCore/Tests/DLLMainTest.cpp:111]: In that case, they are likely calling `AZ::Environment::Attach` which would now be removed, thereby forcing an update on their end anyways.

File: Code/Framework/AzCore/AzCore/O3DEKernelConfiguration.h [hunk 6]
```diff
@@ -0,0 +1,21 @@
+#pragma once
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/AzCore/O3DEKernelConfiguration.h:1]: The `#pragma once` is at the wrong location within the file.

It should be after the copyright.

File: Code/Framework/AzCore/CMakeLists.txt [hunk 7]
```diff
@@ -20,6 +20,20 @@ if(PAL_TRAIT_PROF_PIX_SUPPORTED)
...
+        AzCore/o3dekernel_files.cmake
+    PLATFORM_INCLUDE_FILES
+        ${pal_dir}/o3dekernel_${PAL_PLATFORM_NAME_LOWERCASE}.cmake
+    INCLUDE_DIRECTORIES
+        PUBLIC
+            .
+            ${pal_dir}
+            ${common_dir}
+)
+
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/CMakeLists.txt:36]: Are there plans to be moving the O3DEKernel library to it's own folder structure?

File: Code/Framework/AzCore/AzCore/o3dekernel_files.cmake [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+set(FILES
+    Module/Environment.cpp
+    Module/Environment.h
+)
+
+if(NOT LY_MONOLITHIC_GAME)
+    list(APPEND FILES 
+        std/hash.cpp
+    )
+endif()
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/AzCore/o3dekernel_files.cmake:18]: We could just take the `hash_next_bucket_size` function in hash.cpp and move it to hash.h and avoid the need to selectively include a cpp file for the O3DEKernel

File: Code/Framework/AzCore/Tests/DOM/DomPrefixTreeBenchmarks.cpp [hunk 9]
```diff
@@ -59,7 +59,7 @@ namespace AZ::Dom::Benchmark
     BENCHMARK_DEFINE_F(DomPrefixTreeBenchmark, FindValue_ExactPath)(benchmark::State& state)
     {
         SetupTree(state);
-        for (auto _ : state)
+        for ([[maybe_unused]] auto _ : state)
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/Tests/DOM/DomPrefixTreeBenchmarks.cpp:62]: I am assuming you are using VS2022 and these were flagged as warnings?

Because this change is unrelated to the AZ::Environment work, though I am OK with submitting these changes as a rider bill

File: cmake/Platform/Common/Clang/Configurations_clang.cmake [hunk 10]
```diff
@@ -54,6 +54,7 @@ ly_append_configurations_options(
         -fstack-check
 
     COMPILATION_RELEASE
+        -g
```
Comment (lumberyard-employee-dm) [inline cmake/Platform/Common/Clang/Configurations_clang.cmake:57]: This needs to be removed.
Release shouldn't have debug information in it

File: cmake/Platform/Common/MSVC/Configurations_msvc.cmake [hunk 11]
```diff
@@ -85,6 +85,9 @@ ly_append_configurations_options(
         /Ot             # Favor fast code over small code
         /Oi             # Use Intrinsic Functions
         /Oy             # Omit the frame pointer
+        /Zi
+        /MD
+        /DEBUG
```
Comment (lumberyard-employee-dm) [inline cmake/Platform/Common/MSVC/Configurations_msvc.cmake:90]: The changes to this file should be reset to the previous commit.
This is adding debug information to the MSVC builds

**Review by lumberyard-employee-dm — CHANGES_REQUESTED** [review]

File: Code/Framework/AzCore/Tests/DOM/DomPrefixTreeBenchmarks.cpp (hunk 9 above)
Comment (burelc-amzn) [reply Code/Framework/AzCore/Tests/DOM/DomPrefixTreeBenchmarks.cpp:62]: No, I built with Clang on Windows because I like its diagnostic messages better (and to prevent constant switching between Windows and Linux for the initial development).

File: cmake/Platform/Common/Clang/Configurations_clang.cmake (hunk 10 above)
Comment (burelc-amzn) [reply cmake/Platform/Common/Clang/Configurations_clang.cmake:57]: Will fix.

File: Code/Framework/AzCore/Tests/DLLMainTest.cpp (hunk 5 above)
Comment (burelc-amzn) [reply Code/Framework/AzCore/Tests/DLLMainTest.cpp]: Fixed.

**Review by AMZN-alexpete — APPROVED** [review]

**Review by lumberyard-employee-dm — APPROVED** [review]

**Review by amzn-sj — APPROVED** [review]

**Review by lawsonamzn — APPROVED** [review]
LGTM.  Highlights how much extra aggro caused by the SceneCore/SceneData thing not being a normal DLL or not being a normal AZ module.

**Review by Kadino — APPROVED** [review]
Changes to AzTest are a nice simplification.
